dan_layer/consensus/src/hotstuff/on_receive_vote.rs

```diff
@@ -1,109 +1,142 @@
-        //             .collect::<Vec<_>>();
-        //
-        //         let merged_proof = MergedBalancedBinaryMerkleProof::create_from_proofs(proofs).unwrap();
-        //         let leaf_hashes = votes.iter().map(|v| v.node_hash()).collect::<Vec<_>>();
-        //
-        //         // TODO: Check all votes
-        //         let main_vote = votes.get(0).unwrap();
-        //
-        //         let qc = QuorumCertificate::new(
-        //             node.payload_id(),
-        //             node.payload_height(),
-        //             *node.hash(),
-        //             node.height(),
-        //             node.shard(),
-        //             node.epoch(),
-        //             self.public_key.clone(),
-        //             main_vote.decision(),
-        //             main_vote.all_shard_pledges().clone(),
-        //             validator_metadata,
-        //             Some(merged_proof),
-        //             leaf_hashes,
-        //         );
-        //         self.update_high_qc(&mut tx, node.proposed_by().clone(), qc)?;
-        //
-        //         on_propose = Some((node.shard(), node.payload_id()));
-        //     }
-        //
-        //     // commit the transaction
-        //     tx.commit()?;
-        // }
-        //
-        // // Propose the next node
-        // if let Some((shard_id, payload_id)) = on_propose {
-        //     // TODO: This should go in a some component that controls message flows and events
-        //     let epoch = self.epoch_manager.current_epoch().await?;
-        //     let committee = self.epoch_manager.get_committee(epoch, shard_id).await?;
-        //     if committee.is_empty() {
-        //         return Err(HotStuffError::NoCommitteeForShard { shard: shard_id, epoch });
-        //     }
-        //     if self.is_leader(payload_id, shard_id, &committee)? {
-        //         self.leader_on_propose(shard_id, payload_id).await?;
-        //     }
-        // }
-        // Ok(())
+
+        let addr = self.epoch_manager.get_our_validator_addr(message.epoch).await?;
+        let committee = self.epoch_manager.get_local_committee(message.epoch).await?;
+        if !committee.contains(&addr) {
+            return Err(HotStuffError::ReceivedMessageFromNonCommitteeMember {
+                epoch: message.epoch,
+                sender: from.to_string(),
+                context: "OnVoteReceived".to_string(),
+            });
+        }
+
+        if !self.leader_strategy.is_leader(&addr, &committee, &message.block_id, 0) {
+            return Err(HotStuffError::NotTheLeader {
+                details: format!("Not this leader for block {}, vote sent by {}", message.block_id, addr),
+            });
+        }
+
+        let local_commmittee_shard = self.epoch_manager.get_local_committee_shard(message.epoch).await?;
+
+        // Get the sender shard, and check that they are in the local committee
+        let sender_shard = self
```

Comment:
Seems to be a mismatch of committees and shards. Is shard the shard space or the individual state address?

Reply:
This returns a ShardId, it is the ShardId/Key of the sender. The `CommitteeShard` struct is all the info we need to determine if a shard is in the shard space. I think we need to find some ubiquitous language for these concepts. 

Currently I have:
CommitteeShard = a "slice" of the 256-bit shard space. 
ShardId = a single shard id, a single u256 number. This ID applies to substates and validators
Bucket = (ShardBucket?) is the index of the CommitteeShard in the 256-bit shard space.